lib/seahorn/BvOpSem2RawMemMgr.hh

```diff
@@ -22,71 +24,141 @@ class RawMemManager : public OpSemMemManager {
+      assert(!strct::isStructVal(raw_val));
+      m_v = raw_val;
+    }
+
+    Expr v() const { return m_v; }
+    Expr toExpr() const { return v(); }
+    explicit operator Expr() const { return toExpr(); }
+
+    Expr getRaw() { return m_v; }
+  };
+
+  struct PtrTyImpl {
+    Expr p_v;
+
+    explicit PtrTyImpl(Expr &&raw_val) {
+      assert(!strct::isStructVal(raw_val));
+      p_v = std::move(raw_val);
+    }
+
+    explicit PtrTyImpl(const Expr &raw_val) {
+      assert(!strct::isStructVal(raw_val));
+      p_v = raw_val;
+    }
+
+    Expr v() const { return p_v; }
+    Expr toExpr() const { return v(); }
+    explicit operator Expr() const { return toExpr(); }
+
+    Expr getRaw() { return p_v; }
+  };
+
+  struct MemSortTyImpl {
+    Expr m_mem_sort;
+
+    explicit MemSortTyImpl(Expr &&mem_sort) {
+      m_mem_sort = std::move(mem_sort);
+    }
+
+    explicit MemSortTyImpl(const Expr &mem_sort) { m_mem_sort = mem_sort; }
+
+    Expr v() const { return m_mem_sort; }
+    Expr toExpr() const { return v(); }
+    explicit operator Expr() const { return toExpr(); }
+  };
+
+  struct PtrSortTyImpl {
+    Expr m_ptr_sort;
+
+    explicit PtrSortTyImpl(Expr &&ptr_sort) {
+      m_ptr_sort = std::move(ptr_sort);
+    }
+
+    explicit PtrSortTyImpl(const Expr &ptr_sort) { m_ptr_sort = ptr_sort; }
+
+    Expr v() const { return m_ptr_sort; }
+    Expr toExpr() const { return v(); }
+    explicit operator Expr() const { return toExpr(); }
+  };
+
   /// Right now everything is an expression. In the future, we might have
   /// other types for PtrTy, such as a tuple of expressions
-  using PtrTy = OpSemMemManager::PtrTy;
-  using MemValTy = OpSemMemManager::MemValTy;
-  using PtrSortTy = OpSemMemManager::PtrSortTy;
-  using MemSortTy = OpSemMemManager::MemSortTy;
-  using MemRegTy = OpSemMemManager::MemRegTy;
+  using PtrTy = PtrTyImpl;
+  using MemValTy = MemValTyImpl;
+  using PtrSortTy = PtrSortTyImpl;
+  using MemSortTy = MemSortTyImpl;
+  using MemRegTy = Expr;
   // setting TrackingTag to int disqualifies this class as having tracking
   using TrackingTag = int;
 
-  PtrTy ptrSort() const override { return m_ctx.alu().intTy(ptrSzInBits()); }
+  /// \brief A null pointer expression (cache)
+  PtrTy m_nullPtr;
+
+  /// \brief Register that contains the value of the stack pointer on
+  /// function entry
+  PtrTy m_sp0;
+
+  PtrSortTy ptrSort() const {
+    return PtrSortTy(m_ctx.alu().intTy(ptrSizeInBits()));
+  }
 
   /// \brief Allocates memory on the stack and returns a pointer to it
   /// \param align is requested alignment. If 0, default alignment is used
-  PtrTy salloc(unsigned bytes, uint32_t align = 0) override;
+  PtrTy salloc(unsigned bytes, uint32_t align = 0);
 
   /// \brief Allocates memory on the stack and returns a pointer to it
-  PtrTy salloc(Expr elmts, unsigned typeSz, uint32_t align = 0) override;
+  PtrTy salloc(const Expr &elmts, unsigned typeSz, uint32_t align = 0);
```

Comment:
`Expr` are smart pointers and are usually passed by value. This is an outlier.

Reply:
got it.Fixed.